### PR TITLE
Support `JsonEnum.valueField` being set with `'index'`

### PR DESCRIPTION
**Pull request by kevmoo**
Fixes https://github.com/google/json_serializable.dart/issues/1249

### CODE REVIEW

**Comment by kevmoo** [comment]
Feel free to squash/merge and delete the branch if you're happy

File: json_serializable/lib/src/enum_utils.dart [hunk 1]
```diff
@@ -88,6 +88,20 @@ Object? _generateEntry({
...
       final e = fieldElementType.getField(valueField);
 
+      if (e == null && valueField == 'index') {
+        final values = fieldElementType.getField('values')!;
+
+        final items = ConstantReader(values.computeConstantValue()).listValue;
+
+        return items
+            .singleWhere((element) {
+              final name = element.getField('_name')!.toStringValue()!;
```
Comment (jakemac53) [inline json_serializable/lib/src/enum_utils.dart:98]: is `_name` here something that is guaranteed to stay the same?

Comment (scheglov) [reply json_serializable/lib/src/enum_utils.dart:98]: I'm not sure what you are trying to achieve here. You have `FieldElement field` which is a constant in an `EnumElement`. Then you try to find the corresponding `DartObject` value in its fields, using the name, and get back its index in `values`. If I understand it correctly, then why not use `enumElement.fields.where((field) => field.isEnumConstant).toList().indexOf(fieldElement)`?

Comment (kevmoo) [reply json_serializable/lib/src/enum_utils.dart]: Thanks! done!

**Review by jakemac53 — APPROVED** [review]

**Pull request merged**